numpyro/infer/mcmc.py

```diff
@@ -615,6 +621,257 @@ def __init__(self,
+
+        locs, scales = _get_proposal_loc_and_scale(zs, loc, inv_mass_matrix_sqrt, z)
+        zs_ = np.concatenate([z[None, :], zs])
+        pes_ = np.concatenate([pe[None], pes])
+        if inv_mass_matrix_sqrt.ndim == 2:  # dense_mass
+            log_weights = dist.MultivariateNormal(locs, scale_tril=scales).log_prob(zs_) + pes_
+        else:
+            log_weights = dist.Normal(locs, scales).log_prob(zs_).sum(-1) + pes_
+        log_weights = np.where(np.isnan(log_weights), -np.inf, log_weights)
+        j = categorical_logits(rng_key_reject, log_weights)
+        # XXX: numpy.delete is not available in JAX
+        zs = np.where(np.arange(zs.shape[0])[:, None] < j, zs_[:-1], zs_[1:])
+        pes = np.where(np.arange(pes.shape[0]) < j, pes_[:-1], pes_[1:])
+        loc = locs[j]
+        inv_mass_matrix_sqrt = scales[j]
+        adapt_state = SAAdaptState(zs, pes, loc, inv_mass_matrix_sqrt)
+
+        # XXX: we make a modification of SA sampler in [1]
+        # in [1], each MCMC state contains N points
+        # here we do resampling to pick randomly a point from N points
+        k = categorical_logits(rng_key_accept, np.zeros(zs.shape[0]))
+        z = unravel_fn(zs[k])
+        potential_energy = pes[k]
+
+        accept_prob = 1 - np.exp(log_weights[0] - logsumexp(log_weights))
+        itr = sa_state.i + 1
+        n = np.where(sa_state.i < wa_steps, itr, itr - wa_steps)
+        mean_accept_prob = sa_state.mean_accept_prob + (accept_prob - sa_state.mean_accept_prob) / n
+        return SAState(itr, z, potential_energy, accept_prob, mean_accept_prob,
+                       diverging, adapt_state, rng_key)
+
+    return init_kernel, sample_kernel
+
+
+# TODO: this shares almost the same code as HMC, so we can abstract out much of the implementation
+class SA(MCMCKernel):
+    """
+    Sample Adaptive MCMC, a gradient-free sampler.
+
+    This is a very fast (in term of n_eff / s) sampler but requires
+    many warmup (burn-in) steps. In each MCMC step, we only need to
+    evaluate potential function at one point.
+
+    **References:**
+
+    1. *Sample Adaptive MCMC* (https://papers.nips.cc/paper/9107-sample-adaptive-mcmc),
+       Michael Zhu
+
+    :param model: Python callable containing Pyro :mod:`~numpyro.primitives`.
+        If model is provided, `potential_fn` will be inferred using the model.
+    :param potential_fn: Python callable that computes the potential energy
+        given input parameters. The input parameters to `potential_fn` can be
+        any python collection type, provided that `init_params` argument to
+        `init_kernel` has the same type.
+    :param int adapt_state_size: The number of points to generate proposal
+        distribution. Defaults to 2 times latent size.
+    :param bool dense_mass:  A flag to decide if mass matrix is dense or
+        diagonal (default to ``dense_mass=True``)
+    :param callable init_strategy: a per-site initialization function.
```

Comment:
can you describe how you're setting the N initial samples?

Reply:
Makes sense. We can include this in both HMC as well as this class, though I'm not sure how useful it will be to be able to customize this.

Reply:
Me too. Let's wait for some user requests before adding it, or if we find setting initial inverse_mass_matrix is useful in our experiments.